Add Electron Hub as Chat Completions Provider

snowby666 opened the pull request:
<!-- Put X in the box below to confirm -->

## Checklist:

- [x] I have read the [Contributing guidelines](https://github.com/SillyTavern/SillyTavern/blob/release/CONTRIBUTING.md).

snowby666 committed:
Supported max tokens + fixed wrong image model mapping

Cohee1207 commented on public/index.html:
```diff
@@ -3466,6 +3467,20 @@ <h4 data-i18n="Groq Model">Groq Model</h4>
                                 <option value="mistral-saba-24b">mistral-saba-24b</option>
                             </select>
+                        </div>
+                            <div id="electronhub_form" data-source="electronhub"></div>
```
This tag is incorrectly nested

Cohee1207 commented:
What about the following features?

1. Tool calling
2. Prompt image inlining (images in prompts)
3. Images in chat completion responses (Google's latest "nano banana" model)
4. Reasoning settings (e.g. reasoning effort)

snowby666 commented:
> What about the following features?
> 
> 1. Tool calling
> 2. Prompt image inlining (images in prompts)
> 3. Images in chat completion responses (Google's latest "nano banana" model)
> 4. Reasoning settings (e.g. reasoning effort)

I've fixed nesting error and added support for 1, 2 and 4.

For 3, the API will auto handle this and output the edited image as xml tags inline with the response from the models (`gemini-2.0-flash-preview-image-generation` and `gemini-2.5-flash-image-preview`). You can ask follow-up prompts to edit the image as well.


```yaml
<img src="image generated by gemini">
```

<img width="1338" height="1437" alt="image" src="https://github.com/user-attachments/assets/f9656697-fed2-4b24-91af-2c8a7fe5db4a" />

Cohee1207 commented on public/index.html:
```diff
@@ -3467,6 +3468,20 @@ <h4 data-i18n="Groq Model">Groq Model</h4>
+                            <div data-for="api_key_electronhub" class="neutral_warning" data-i18n="For privacy reasons, your API key will be hidden after you click 'Connect'.">
+                                For privacy reasons, your API key will be hidden after you click 'Connect'.
+                            </div>
+                            <h4 data-i18n="ElectronHub Model">Electron Hub Model</h4>
```
data-i18n value doesn't match with the added string to locale jsons

Cohee1207 commented on public/index.html:
```diff
@@ -3467,6 +3468,20 @@ <h4 data-i18n="Groq Model">Groq Model</h4>
                             </select>
                         </div>
+                        <div id="electronhub_form" data-source="electronhub">
+                            <h4 data-i18n="ElectronHub API Key">Electron Hub API Key</h4>
```
data-i18n value doesn't match with the added string to locale jsons

Cohee1207 commented on public/index.html:
```diff
@@ -2813,6 +2813,7 @@ <h4 class="margin0" data-i18n="Chat Completion Source">
                                 <option value="vertexai">Google Vertex AI</option>
                                 <option value="mistralai">MistralAI</option>
                                 <option value="moonshot">Moonshot AI</option>
+                                <option value="electronhub">Electron Hub</option>
```
The list should be sorted alphabetically

Cohee1207 commented on public/scripts/extensions/stable-diffusion/index.js:
```diff
@@ -3564,6 +3599,32 @@ async function generateHuggingFaceImage(prompt, signal) {
+        headers: getRequestHeaders(),
+        signal: signal,
+        body: JSON.stringify({
+            model: extension_settings.sd.model,
```
Does it support any other parameters? Like width / height at least?

snowby666 replied:
We use a list of available sizes instead of width and height. I've add a function to get the closest values.

Cohee1207 commented on public/scripts/extensions/stable-diffusion/settings.html:
```diff
@@ -45,6 +45,7 @@
                 <option value="falai">FAL.AI</option>
                 <option value="google">Google AI</option>
                 <option value="huggingface">HuggingFace Inference API (serverless)</option>
+                <option value="electronhub">Electron Hub</option>
```
The list should be sorted alphabetically

Cohee1207 commented on public/scripts/extensions/stable-diffusion/settings.html:
```diff
@@ -93,6 +94,9 @@
                 <input id="sd_huggingface_model_id" type="text" class="text_pole"  data-i18n="[placeholder]e.g. black-forest-labs/FLUX.1-dev" placeholder="e.g. black-forest-labs/FLUX.1-dev" value="" />
             </div>
+            <div data-sd-source="electronhub">
+                <i>Hint: Save an API key in the Electron Hub API settings to use it here.</i>
```
Consider adding (Chat Completion) after the API name (see NanoGPT example below)

Cohee1207 commented on public/scripts/openai.js:
```diff
@@ -2302,10 +2339,12 @@ async function sendOpenAIRequest(type, messages, signal, { jsonSchema = null } =
         chat_completion_sources.XAI,
         chat_completion_sources.POLLINATIONS,
         chat_completion_sources.AIMLAPI,
+        chat_completion_sources.ELECTRONHUB,
```
The source is added to the list twice. Once is enough.

Cohee1207 commented on public/scripts/openai.js:
```diff
@@ -2172,6 +2200,15 @@ async function sendOpenAIRequest(type, messages, signal, { jsonSchema = null } =
+        delete generate_data.logit_bias;
+        delete generate_data.stop;
+        delete generate_data.logprobs;
+        delete generate_data.frequency_penalty;
```
You have the freq.pen and pres.pen parameters unlocked in the UI (`index.html`), but they are removed here. Also you're not setting Top K in here, JFYI.

Cohee1207 commented on public/scripts/openai.js:
```diff
@@ -2172,6 +2200,15 @@ async function sendOpenAIRequest(type, messages, signal, { jsonSchema = null } =
+        delete generate_data.logprobs;
+        delete generate_data.frequency_penalty;
+        delete generate_data.presence_penalty;
+        delete generate_data.n;
```
N won't be set by default unless `canMultiSwipe` is true (it will always be false here).
Same with logprobs - there's a separate condition that controls its inclusion into the request.

Cohee1207 commented on public/scripts/openai.js:
```diff
@@ -2172,6 +2200,15 @@ async function sendOpenAIRequest(type, messages, signal, { jsonSchema = null } =
     }
 
+    if (isElectronHub) {
+        delete generate_data.logit_bias;
```
Logit bias will not be set here, as it is controlled by the `logitBiasSources` variable.

Cohee1207 commented on public/scripts/openai.js:
```diff
@@ -3432,11 +3471,13 @@ function loadOpenAISettings(data, settings) {
     oai_settings.cohere_model = settings.cohere_model ?? default_settings.cohere_model;
     oai_settings.perplexity_model = settings.perplexity_model ?? default_settings.perplexity_model;
     oai_settings.groq_model = settings.groq_model ?? default_settings.groq_model;
+    oai_settings.electronhub_model = settings.electronhub_model ?? default_settings.electronhub_model;
```
Duplicated line (one more at line 3480).

Cohee1207 commented on public/scripts/openai.js:
```diff
@@ -4736,6 +4789,15 @@ async function onModelChange() {
         oai_settings.cometapi_model = value;
     }
 
+    if ($(this).is('#model_electronhub_select')) {
```
Duplicated code block.

Cohee1207 commented on public/scripts/openai.js:
```diff
@@ -4907,6 +4969,23 @@ async function onModelChange() {
     }
 
+    if (oai_settings.chat_completion_source == chat_completion_sources.ELECTRONHUB) {
+        if (oai_settings.max_context_unlocked) {
```
Consider extracting this logic into a module-scoped function

Cohee1207 commented on public/scripts/openai.js:
```diff
@@ -6312,9 +6409,11 @@ export function initOpenAI() {
     $('#model_cohere_select').on('change', onModelChange);
     $('#model_perplexity_select').on('change', onModelChange);
     $('#model_groq_select').on('change', onModelChange);
+    $('#model_electronhub_select').on('change', onModelChange);
```
Duped line once again.

Cohee1207 commented on src/endpoints/backends/chat-completions.js:
```diff
@@ -1194,6 +1195,109 @@ async function sendAimlapiRequest(request, response) {
+        }
+
+        if (request.body.reasoning_effort) {
+            const reasoningEffortModels = ["o3-mini", "o1", "o3", "o4-mini", "codex-mini-latest", "o3-pro", "gpt-5-mini", "gpt-5-nano", "gpt-5"];
```
Is there a better way? Maybe add a condition somewhere on the frontend in the function that sets the reasoning effort in the request? I'd like not to maintain a hard coded models list.

snowby666 replied:
Only a few OpenAI models use this param so I'm not sure what you mean by this. Perhaps fetch a list of models supported this param from the API?

Cohee1207 replied:
Similarly to `isImageInliningSupported` you can check the `models_list` on frontend before setting the reasoning_effort in `function getReasoningEffort`

Cohee1207 commented on src/endpoints/backends/chat-completions.js:
```diff
@@ -1194,6 +1195,109 @@ async function sendAimlapiRequest(request, response) {
+            'temperature': request.body.temperature,
+            'max_tokens': request.body.max_tokens,
+            'stream': request.body.stream,
+            'presence_penalty': request.body.presence_penalty,
```
presence_penalty and frequency_penalty are already removed on the frontend (see another comment above)

Cohee1207 commented on src/endpoints/backends/chat-completions.js:
```diff
@@ -1194,6 +1195,109 @@ async function sendAimlapiRequest(request, response) {
+            'presence_penalty': request.body.presence_penalty,
+            'frequency_penalty': request.body.frequency_penalty,
+            'top_p': request.body.top_p,
+            'top_k': request.body.top_k,
```
top_k will be undefined since it's not sent from the frontend

Cohee1207 reviewed (changes requested):
Please sync your branch with the latest staging before fixing any of the comments. The diff seems to be out of date with the changes in the branch.

snowby666 committed:
fixed merge conflicts

snowby666 committed:
fixed merge conflicts

snowby666 committed:
fixed merge conflicts

snowby666 committed:
updated the logic

snowby666 committed:
updated the logic

snowby666 committed:
replaced hard coded reasoning_effort model list with a dynamic function

snowby666 committed:
replaced hard coded reasoning_effort mode list with a dynamic function

snowby666 committed:
replaced hard coded reasoning_effort model list with a dynamic function

Cohee1207 committed:
Fix eslint

Cohee1207 committed:
Merge branch 'staging' into electron_hub

Cohee1207 committed:
Adjust reasoning effort logic

Cohee1207 committed:
Code clean-up

Cohee1207 committed:
Merge branch 'staging' into electron_hub

Cohee1207 committed:
Add logo

Cohee1207 committed:
Add inline image quality

Cohee1207 committed:
Fix multimodal models list

Cohee1207 committed:
Fix seed not passed

Cohee1207 committed:
Add "detail" error parser

Cohee1207 commented:
Any idea why this request fails? It works fine with the OpenAI API.

```js
messages: [
    { role: 'system', content: 'Reply as Assistant.' },
    { role: 'user', content: "What's the weather in Rome?" },
    {
      role: 'assistant',
      tool_calls: [
        {
          id: 'call_ 2ltA8VRyrs05rs7Hs9QX1hom,
          type: 'function',
          function: {
            arguments: '{"location":"Rome, Italy","units":"metric","condition":true,"temperature":true}',
            name: 'GetCurrentWeather'
          }
        }
      ]
    },
    {
      role: 'tool',
      content: 'Sunny, 27.5°C',
      tool_call_id: 'call_ 2ltA8VRyrs05rs7Hs9QX1hom'
    }
  ]
  ```
Error returned:
`{"detail":{"error":{"message":"Invalid messages format.","type":"error","param":null,"code":400}}}`

snowby666 commented:
> Any idea why this request fails? It works fine with the OpenAI API.
> 
> ```js
> messages: [
>     { role: 'system', content: 'Reply as Assistant.' },
>     { role: 'user', content: "What's the weather in Rome?" },
>     {
>       role: 'assistant',
>       tool_calls: [
>         {
>           id: 'call_ 2ltA8VRyrs05rs7Hs9QX1hom,
>           type: 'function',
>           function: {
>             arguments: '{"location":"Rome, Italy","units":"metric","condition":true,"temperature":true}',
>             name: 'GetCurrentWeather'
>           }
>         }
>       ]
>     },
>     {
>       role: 'tool',
>       content: 'Sunny, 27.5°C',
>       tool_call_id: 'call_ 2ltA8VRyrs05rs7Hs9QX1hom'
>     }
>   ]
> ```
> 
> Error returned: `{"detail":{"error":{"message":"Invalid messages format.","type":"error","param":null,"code":400}}}`

There was some misconfig when validating the tools arguments. I've made a patch for this.

Cohee1207 reviewed (approved):
LGTM. I also see in the docs that you have TTS and embedding endpoints, feel free to add that into appropriate ST extensions.
